client/verta/verta/runtime.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: utf-8 -*-
+
+import json
+from typing import Any, Dict
+
+
+class LogHandler:
+    """
+    Container class for collecting and dumping log entries.
+    """
+    def __init__(self):
+        self.attributes: Dict[str, Any] = dict()
+
+    def __repr__(self):
+        return str(self.attributes)
+
+    def add(self, **kwargs):
+        # Overwrites existing key-values
+        self.attributes.update(**kwargs)
+
+    def wipe_clean(self):
+        self.attributes = dict()
+
+# Use a single handler object that is wiped clean after each log entry
+# to minimize memory usage
+LOG_HANDLER = LogHandler()
+
+
+def log(**kwargs) -> None:
+    """
+    Log any number of key-value pairs to the current prediction context.
+    Values are deleted upon completion of the active model's predict() method.
+
+    Parameters
+    ----------
+    kwargs : Any
+        Arguments and values to be added as key-value pairs to the current log record.
+        Values must be serializable to JSON.
+
+    Returns
+    -------
+    None
+    """
+    LOG_HANDLER.add(**kwargs)
+
+
+class ContextLogger():
+    """
+    Context manager for aggregating key-value pairs into a custom log entry.
+    Any logged data is deleted upon exiting the context.
+
+    Attributes
+    ----------
+    as_json : str
+        A JSON string representation of all the current saved logging context.
+    """
+    def __init__(self):
+        self.context = LOG_HANDLER
+        if bool(self.context.attributes): # if not empty dict
+            self.context.wipe_clean()
+
+    def log(self, **kwargs) -> None:
```

Comment:
Isn't this supposed to be

```suggestion
    def log(self, key, val) -> None:
```

as per the design doc we agreed on?

https://github.com/VertaAI/docs/blob/main/design-docs/2023-01-model-prediction-logging.md#logging-key-values

I don't see mention of any reason to have deviated from that.

Reply:
I suppose we need to come to an agreement on exactly how prescriptive the design doc is supposed to be.  Using `**kwargs` here allows the user to add any number of key-value pairs they want to log at that moment.

`runtime.log(key1='val1', key2='val2', key3='val3')`
just makes more sense than
```
runtime.log('key1', 'val1')
runtime.log('key2', 'val2')
runtime.log('key3', 'val3')
```
and forces the key's used to be log-friendly, queryable strings that can be legit column names in the final Athena table

Reply:
> I suppose we need to come to an agreement on exactly how prescriptive the design doc is supposed to be.

I'm not opposed in principle to deviating from a design doc, if we have reason to! Buuuuuuuut

> ```python
> runtime.log(key1='val1', key2='val2', key3='val3')
> ```
> just makes more sense

I disagree with this. I think there _is_ merit in an argument that it forces the keys to be strings, but the idea that a user's arbitrary items for logging are _keyword arguments_ to a function is extremely unintuitive to me. Like, I don't think that's what function parameters are meant for, and no other libraries come to mind that do this: allow a user to pass their own arbitrary key-value items as named arguments (though if you've found any examples, I'd be interested in reviewing them).

If we'd like to support

> add any number of key-value pairs they want to log at that moment.

I'd personally recommend just accepting a whole dictionary, like Weights & Biases does (https://docs.wandb.ai/guides/track/log), for which we also have established precedence for in our client via methods like [`log_hyperparameters()`](https://verta.readthedocs.io/en/main/_autogen/verta.tracking.entities.ExperimentRun.html#verta.tracking.entities.ExperimentRun.log_hyperparameters).

Reply:
I think it'd be more prudent to raise this question in #dev-python than try to work it out between just the two of us. If a consensus was reached already during the design doc phase, I'm concerned about the implications of making last-mile changes without a similar breadth of rigor.

Reply:
Refactored to accept separate `key` and `value` parameters and simply update the current context dictionary with that k-v pair.

Reply:
If we find that we need to enforce some constraints on the "key" parameter later, we can always add that in another PR.

Reply:
To be clear, I would've been good with raising this as a discussion point in one of our Slack channels. My personal disagreement and recommendation aren't absolute.

Reply:
Understood, and thanks. It's not so much that I don't feel comfortable challenging, but that you have a breadth of valuable experience with this product, client code, and certain Python constructs upon which I depend.  Very often we bounce back and forth, and I wind up convinced that you are right :) 

I honestly didn't feel that strongly about it and there is nothing about doing it with `kwargs` that I think would have a significant positive impact on the overall user experience.